src/foundations/type/Type.tsx

```diff
@@ -0,0 +1,124 @@
+import styled, { css } from 'styled-components';
+import { FlattenSimpleInterpolation } from 'styled-components/ts3.6';
+
+import { Utils } from '../../utils';
+
+import { BODY_SIZES, DEFAULT_TYPE_SIZE, HEADING_SIZES, HEADING_ELEMENTS } from './Type.constants';
+import { PType } from './Type.props';
+import { TTypeSize, TTypeType } from './Type.types';
+
+function generateFontStyleVariations(
+    typeSize: TTypeSize,
+    fontType: TTypeType
+): FlattenSimpleInterpolation {
+    return css`
+        &[data-size='${typeSize}'] {
+            font-size: var(--${fontType}-font-size-${typeSize});
+            line-height: var(--${fontType}-line-height-${typeSize});
+
+            &[data-margin='none'] {
+                margin: 0;
+            }
+            &[data-margin='top'] {
+                margin: var(--${fontType}-margin-top-${typeSize}) 0 0;
+            }
+            &[data-margin='bottom'] {
+                margin: 0 0 var(--${fontType}-margin-bottom-${typeSize});
+            }
+            &[data-margin='both'] {
+                margin: var(--${fontType}-margin-top-${typeSize}) 0
+                    var(--${fontType}-margin-bottom-${typeSize});
+            }
+        }
+    `;
+}
+
+const Type = styled.p
+    .attrs((props: PType) => ({
+        // it is possible to remap props, so we do not need to pass down the
+        // `as` property from styled-components and prevent usage of
+        // unsupported HTML tags
+        as: props.element,
+        'data-size': props.size || DEFAULT_TYPE_SIZE,
+        'data-type': HEADING_ELEMENTS.includes(props.element) ? 'heading' : 'body',
+        'data-weight': props.weight,
+        'data-margin': props.margin,
+        'data-color': props.color || 'primary',
+    }))
+    .withConfig({
+        shouldForwardProp: Utils.doNotForwardProperties([
+            'size',
+            'type',
+            'weight',
+            'element',
+            'margin',
+        ]),
+    })<PType>`
+        // set body type styles (default)
+        // - defaults
+        font-family: var(--body-font-family);
+        font-size: var(--base-font-size);
+        font-weight: var(--font-weight-regular);
+        line-height: var(--base-line-height);
+        margin: var(--body-margin-top-${DEFAULT_TYPE_SIZE}) 0
+            var(--body-margin-bottom-${DEFAULT_TYPE_SIZE});
```

Comment:
Possibly not really necessary, but setting the margin defaults here with the rest of the defaults for consistency.